From 1663ae0139040b5d9daaa67a0592b75def489e5b Mon Sep 17 00:00:00 2001
From: Konstantin Porotchkin <kostap@marvell.com>
Date: Tue, 17 Dec 2019 15:56:20 +0200
Subject: [PATCH 1039/1239] drivers/pci_endpoint: fix mvebu_ep_uio bar setups

- take the bar physical address into account
- fix the bar mask for disabling unused regions

Change-Id: Ib63f02cdd811cb0552208e1c039229a536d0a42b
Signed-off-by: Konstantin Porotchkin <kostap@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/20450
Reviewed-by: Stefan Chulski <stefanc@marvell.com>
Tested-by: sa_ip-sw-jenkins <sa_ip-sw-jenkins@marvell.com>
---
 drivers/pci_endpoint/pcie_mvebu_ep_uio.c | 13 ++++++++-----
 1 file changed, 8 insertions(+), 5 deletions(-)

diff --git a/drivers/pci_endpoint/pcie_mvebu_ep_uio.c b/drivers/pci_endpoint/pcie_mvebu_ep_uio.c
index 4330ce9db7..bd06317d75 100644
--- a/drivers/pci_endpoint/pcie_mvebu_ep_uio.c
+++ b/drivers/pci_endpoint/pcie_mvebu_ep_uio.c
@@ -29,7 +29,9 @@ struct uio_pci {
 };
 
 /* make sure we have at least one mem regions to map the host ram */
-#define MAX_BAR_MAP	4
+#define MAX_BAR_MAP		4
+#define PCIE_EP_ALL_BARS	0x3f
+
 
 static int uio_pci_ep_probe(struct udevice *dev)
 {
@@ -93,6 +95,7 @@ static int uio_pci_ep_probe(struct udevice *dev)
 		bar_mask |= 1 << bar_id;
 
 		bar.barno = bar_id;
+		bar.phys_addr = res.start;
 		bar.size = resource_size(&res);
 		if (!is_power_of_2(bar.size)) {
 			printf("BAR-%d size in not power of 2\n", bar_id);
@@ -111,10 +114,10 @@ static int uio_pci_ep_probe(struct udevice *dev)
 		}
 	}
 
-	for (bar_id = 0; bar_id < MAX_BAR_MAP; bar_id++) {
-		if (bar_mask & (1 << bar_id))
-			continue;
-		pci_ep_clear_bar(uio_pci->ep, 0, bar_id);
+	bar_mask = PCIE_EP_ALL_BARS & ~bar_mask;
+	for (bar_id = 0; bar_mask >>= 1; bar_id++) {
+		if (bar_mask & 1)
+			pci_ep_clear_bar(uio_pci->ep, 0, bar_id);
 	}
 
 	/* remap host RAM to local memory space  using shift mapping.
-- 
2.29.0

